lib/srv/desktop/deskproto/proto.go

```diff
@@ -0,0 +1,254 @@
+// Package deskproto implements the desktop protocol encoder/decoder.
+// See https://github.com/gravitational/teleport/blob/master/rfd/0037-desktop-access-protocol.md
+//
+// TODO(awly): complete the implementation of all messages, even if we don't
+// use them yet.
+package deskproto
+
+import (
+	"bytes"
+	"encoding/binary"
+	"image"
+	"image/png"
+	"io"
+
+	"github.com/gravitational/trace"
+)
+
+// MessageType identifies the type of the message.
+type MessageType byte
+
+const (
+	TypeClientScreenSpec = MessageType(1)
+	TypePNGFrame         = MessageType(2)
+	TypeMouseMove        = MessageType(3)
+	TypeMouseButton      = MessageType(4)
+	TypeKeyboardButton   = MessageType(5)
+	TypeClientUsername   = MessageType(7)
+)
+
+// Message is a Go representation of a desktop protocol message.
+type Message interface {
+	Encode() ([]byte, error)
+}
+
+// Decode decodes the wire representation of a message.
+func Decode(buf []byte) (Message, error) {
+	if len(buf) == 0 {
+		return nil, trace.BadParameter("input desktop protocol message is empty")
+	}
+	switch buf[0] {
+	case byte(TypeClientScreenSpec):
```

Comment:
Nit: If you did `switch MessageType(buf[0])` instead, you wouldn't need byte casts in each case branch.